idaes/apps/uncertainty_propagation/uncertainties.py

```diff
@@ -0,0 +1,511 @@
+    
+    return results
+
+def propagate_uncertainty(model_uncertain, theta, cov, theta_names, tee=False, solver_options=None):
+    """This function calculates gradient vector, expectation, and variance of the objective function and constraints
+    of the model for given estimated optimal parameters and covariance matrix of parameters. It calculates 
+    error propagation of the objective function and constraints by using gradient vector and covariance matrix. 
+    
+    Parameters
+    ----------
+    model_uncertain: function or Pyomo ConcreteModel
+        function is a a python Function that generates an instance of the Pyomo model
+    theta: dict
+        Estimated parameters 
+    cov: numpy.ndarray
+        Covariance matrix of parameters 
+    theta_names: list of strings
+        List of estimated Var names
+    var_dic: dictionary
+        If any original variable contains "'", need an auxiliary dictionary  with keys theta_namess without "'", values with "'".
+        e.g) var_dic: {'fs.properties.tau[benzene,toluene]': "fs.properties.tau['benzene','toluene']", 'fs.properties.tau[toluene,benzene]': "fs.properties.tau['toluene','benzene']"} 
+    tee: bool, optional
+        Indicates that ef solver output should be teed
+    solver_options: dict, optional
+        Provides options to the solver (also the name of an attribute)
+
+    Returns
+    -------
+    gradient_f_dic: dic
+        gradient of the objective function with respect to all decision variables at the optimal solution 
+        with variable name as key e.g) dic = {d(f)/d(x1):0.1, d(f)/d(x2):0.1}
+    gradient_c_dic: dic
+        gradient of the constraints with respect to all decision variables at the optimal solution
+        with constraint number and variable name as key e.g) dic = {d(c1)/d(x1):1.1, d(c4)/d(x2):0.1}
+        Only non-zero gradients are included.
+    dsdp_dic: dict
+        gradient vector of the (decision variables, parameters) with respect to paramerters (=theta_name).    
+        e.g) dict = {'d(x1)/d(p1)': 1.0, 'd(x2)/d(p1)': 0.0, 'd(p1)/d(p1)': 1.0, 'd(p2)/d(p1)': 0.0, 
+                     'd(x1)/d(p2)': 0.0, 'd(x2)/d(p2)': 1.0, 'd(p1)/d(p2)': 0.0, 'd(p2)/d(p2)': 1.0}
+    propagation_f: dictionary
+        error propagation in the objective function with the dictionary key, 'objective' 
+        if the objective function doesn't include any uncertain parameters, return an empty dictionary        
+    propagation_c: dictionary
+        error propagation in the constraints with the dictionary keys, 'constraints r'
+        where r is the line number. 
+        if no constraint includes uncertain parameters, return an empty dictionary      
+
+    Raises
+    ------
+    Exception
+        When model_uncertain is neither 'ConcreteModel' nor 'function'.
+    """
+    # define Pyomo model
+    if type(model_uncertain).__name__ == 'ConcreteModel':
+        model = model_uncertain
+    elif type(model_uncertain).__name__ == 'function':
```

Comment:
Can import type module(?)

Reply:
@dangunter @Robbybp Any recommendations for a better way to do this?

Reply:
Please try the suggestion from @Robbybp. See the previous comment.